link/src/test/java/com/stripe/android/link/LinkForegroundActivityTest.kt

```diff
@@ -0,0 +1,184 @@
+package com.stripe.android.link
+
+import android.app.Activity
+import android.content.ActivityNotFoundException
+import android.content.Intent
+import androidx.core.net.toUri
+import androidx.lifecycle.Lifecycle
+import androidx.test.core.app.ActivityScenario
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.espresso.intent.Intents.assertNoUnverifiedIntents
+import androidx.test.espresso.intent.Intents.intended
+import androidx.test.espresso.intent.Intents.intending
+import androidx.test.espresso.intent.Intents.times
+import androidx.test.espresso.intent.matcher.IntentMatchers.anyIntent
+import androidx.test.espresso.intent.matcher.IntentMatchers.hasAction
+import androidx.test.espresso.intent.matcher.IntentMatchers.hasData
+import androidx.test.espresso.intent.rule.IntentsRule
+import com.google.common.truth.Truth.assertThat
+import org.hamcrest.Matchers.allOf
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.Robolectric
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.Shadows
+import org.robolectric.shadows.ShadowActivity
+
+@RunWith(RobolectricTestRunner::class)
+internal class LinkForegroundActivityTest {
+    private val popupUrl = "https://checkout.link.com/from_unit_tests"
+    private val redirectUrl = "stripesdk://link_return_url/example.app?payload=foo".toUri()
+
+    @get:Rule
+    val intentsTestRule = IntentsRule()
+
+    @Test
+    fun `finishes with a cancelled result when no popupUrl is passed`() {
+        val intent =
+            Intent(ApplicationProvider.getApplicationContext(), LinkForegroundActivity::class.java)
+
+        val scenario = ActivityScenario.launchActivityForResult<LinkForegroundActivity>(intent)
+        assertThat(scenario.result.resultCode)
+            .isEqualTo(Activity.RESULT_CANCELED)
+        assertNoUnverifiedIntents()
+    }
+
+    @Test
+    fun `finishes with a failure result when activity fails to launch`() {
```

Comment:
Should we also have a test for:

```
override fun onResume() {
        super.onResume()

        if (!isFinishing) {
            if (hasLaunchedPopup) {
                setResult(Activity.RESULT_CANCELED)
                finish()
            } else {
                launchPopup()
            }
        }
    }
```

If launched, then cancel?

Reply:
That's a few tests down! `closing the browser emits a cancelled result`